velox/expression/tests/ExpressionFuzzer.cpp

```diff
@@ -1045,20 +1062,108 @@ core::TypedExprPtr ExpressionFuzzer::generateExpression(
   if (funcIt == funcArgOverrides_.end()) {
     return generateArgs(callable);
   }
   return funcIt->second(callable);
 }
 
+TypePtr ExpressionFuzzer::getConstrainedOutputType(
+    const std::vector<core::TypedExprPtr>& args,
+    const exec::FunctionSignature* signature) {
+  if (signature == nullptr) {
+    return nullptr;
+  }
+  // Checks if any variable is integer constrained, and get the decimal name
+  // style.
+  bool integerConstrained = false;
+  char decimalNameStyle = 0;
+  for (const auto& [variableName, variableInfo] : signature->variables()) {
+    if (variableInfo.isIntegerParameter()) {
+      // If constraints are empty, the integer variable is also regarded to be
+      // constrained as variables are shared across argument and return types.
+      integerConstrained = true;
+      if (variableName == "precision" || variableName == "scale") {
+        decimalNameStyle = 'a';
+        break;
+      }
+      if (variableName.find("precision") != std::string::npos ||
+          variableName.find("scale") != std::string::npos) {
+        decimalNameStyle = 'b';
+        break;
+      }
+      if (variableName.find("i") != std::string::npos) {
+        decimalNameStyle = 'c';
+        break;
+      }
+    }
+  }
+
+  // To handle the constraints between input types and output types of a decimal
+  // function, extracts the input precisions and scales from decimal arguments,
+  // and bind them to integer variables.
+  std::unordered_map<std::string, int> decimalVariablesBindings;
+  column_index_t decimalColIndex = 1;
+  for (column_index_t i = 0; i < args.size(); ++i) {
+    const auto argType = args[i]->type();
+    if (argType->isDecimal()) {
+      const auto [p, s] = getDecimalPrecisionScale(*argType);
+      switch (decimalNameStyle) {
+        case 'a': {
+          decimalVariablesBindings["precision"] = p;
+          decimalVariablesBindings["scale"] = s;
+          break;
+        }
+        case 'b': {
+          const auto column = std::string(1, 'a' + i);
+          decimalVariablesBindings[column + "_precision"] = p;
+          decimalVariablesBindings[column + "_scale"] = s;
+          break;
+        }
+        case 'c': {
+          decimalVariablesBindings["i" + std::to_string(decimalColIndex)] = p;
+          decimalVariablesBindings
+              ["i" + std::to_string(decimalColIndex + kIntegerPairSize)] = s;
+          decimalColIndex++;
+          break;
+        }
+        default:
+          VELOX_UNSUPPORTED(
+              "Unsupported decimal name style {}.", decimalNameStyle);
+      }
+    }
+  }
+
+  // Calculates the matched return type through the argument types with argument
+  // type fuzzer, which evaluates the constraints internally.
+  if (integerConstrained && decimalVariablesBindings.size() > 0 && signature) {
+    ArgumentTypeFuzzer fuzzer{*signature, rng_, decimalVariablesBindings};
+    return fuzzer.fuzzReturnType();
+  }
+  return nullptr;
```

Comment:
I feel like passing `decimalVariablesBindings` to the argument fuzzer exposes the internal implementation details via it constructor. Would it instead make sense to pass `signature->variables()` to the constructor and move all the logic about extracting the `decimalVariablesBindings` there. This would be a natural extension to the constructor similar to the one where we already provide the return type and want to bind the arguments based off of it. Here, we would be providing the already bound variables and would like to get the return type based off of it. what do you think?

Reply:
Thank you for the suggestion. The idea makes sense to me. Since signature is passed to the constructor, `signature->variables()` is not needed. Passed argument types to the constructor.